File: 삼성 코테 기출/14503 로봇 청소기.py
n, m = map(int, input().split())
r, c, d = map(int, input().split())
arr = []

for i in range(n) :
  arr.append(list(map(int, input().split())))

# 벽은 2로 바꾸기
for i in range(n) :
  arr[i] = [2 if n==1 else 0 for n in arr[i]]

# 왼쪽으로 도는 코드
def func(r, c, d) :
  if d == 0 :
    return (r, c-1, 3)
  if d == 3 :
    return (r+1, c, 2)
  if d == 2 :
    return (r, c+1, 1)
  if d == 1 :
    return (r-1, c, 0)

# 뒤로 후진하는 코드
def back(r, c, d) :
  if d == 0 :
    return (r+1, c, 0)
  if d == 3 :
    return (r, c+1, 3)
  if d == 2 :
    return (r-1, c, 2)
  if d == 1 :
    return (r, c-1, 1)

flag = 0
colored = 0
cnt = 0
ans = 0
    
# 종료조건 : 뒤에 벽이 있고(flag = 1) 사방이 모두 칠해져 있을때(cnt = 4)
while not (flag == 1 and cnt == 4) :

  cnt = 0 # 회전 횟수 초기화
  colored = 0
  
  # 3. 비어있지 않거나 벽이면 왼쪽으로 돈다 (d는 바뀐채로 while문 돌기)
  while cnt < 4 :
    # 0. 현재가 칠해져 있지 않다면 칠하고 break
    if arr[r][c] == 0 :
      arr[r][c] = 1
      ans += 1
      break
      
    # 1. 현재가 칠해져 있거나 벽이라면 현재 있는 방향에서 왼쪽을 확인한다
    nx, ny, d = func(r, c, d)
    cnt += 1
    # 1-1. 범위를 벗어난다면 break
    if nx<0 or ny<0 or nx>=n or ny>=m : break
    # 2. 왼쪽이 비어있다면 칠하고 r, c, d 초기화
    if arr[nx][ny] == 0 :
      arr[nx][ny] = 1
      ans += 1
      r, c, d = nx, ny, d
      colored = 1
      break
    
  # 4. 모든 방향에 칠할 수 없다면(cnt=4) 뒤로 1칸 후진하여 루프를 돈다
  if cnt == 4 and colored == 0 :
    r, c, d = back(r, c, d)
    # 5. 뒤가 벽이라면 종료
    if arr[r][c] == 2 :
      flag = 1


print(ans)

File: 삼성 코테 기출/14503 로봇청소기.cpp
#include <iostream>
using namespace std;

#define MAX 50
int N, M; //세로, 가로
int map[MAX + 2][MAX + 2];
int ans;

struct STATUS { int r; int c; int d; };
STATUS st;

const int dc[] = { 0, 1, 0, -1 }; //전진
const int dr[] = { -1, 0, 1, 0 };
const int ddc[] = { 0, -1, 0, 1 }; //후진
const int ddr[] = { 1, 0, -1, 0 };


void input()
{
	cin >> N >> M;
	cin >> st.r >> st.c >> st.d;
	for (int r = 0; r < N; r++) {
		for (int c = 0; c < M; c++) {
			cin >> map[r][c];
		}
	}
}

int solve()
{
	//초기값
	int r = st.r; int c = st.c; int d = st.d;
	int ans = 0;

	while (true) {

		int flag = 0;
		bool clean = false;

		//0. 현재 장소를 청소할 수 있다면 한다
		if (map[r][c] == 0) {
			map[r][c] = 2;
			ans++;
		}

		//1. 현재 방향부터 왼쪽으로 탐색 진행
		for (int i = 0; i < 4; i++) {
			//1-0. 왼쪽 방향
			d = (d + 3) % 4;
			int nc = c + dc[d];
			int nr = r + dr[d];

			//1-1. 좌표 유효성 및 청소 유효성
			if (nc < 0 || nr < 0 || nc >= M || nr >= N) { flag++; continue; }
			if (map[nr][nc] == 1 || map[nr][nc] == 2) { flag++; continue; }

			//1-2. 청소 가능하다면 청소하고 아래 전부 무시
			r = nr; c = nc;
			clean = true;
			break;
		}

		//2. 4방향 모두 불가하고 한번도 청소되지 않았다면 (nr, nc가 0일리가 없음)
		if (clean == false && flag == 4) {
			int nc = c + ddc[d];
			int nr = r + ddr[d];

			//2-1. 뒤가 벽이면 종료
			if (map[nr][nc] == 1 || nc < 0 || nr < 0 || nc >= M || nr >= N) return ans;

			//2-2. 아니면 한 칸 후진
			r = nr; c = nc;
		}
	}
}


int main(void)
{
	input();
	int ans = solve();
	cout << ans;
	return 0;
}